docs/source/export.ir_spec.rst

```diff
@@ -0,0 +1,501 @@
+TensorMeta is never dynamic. The dtype field represents the dtype of the
+output of that node. There are no implicit type promotions in Edge IR. There
+are no strides in FakeTensor.
+
+In other words:
+
+- If the operator in node.target returns a Tensor, then ``node.meta['val']`` is a
+  FakeTensor describing that tensor.
+- If the operator in node.target returns an n-tuple of Tensors, then
+  ``node.meta['val']`` is an n-tuple of FakeTensors describing each tensor.
+- If the operator in node.target returns an int/float/scalar that is known at
+  compile time, then ``node.meta['val']`` is None.
+- If the operator in node.target returns an int/float/scalar that is not known
+  at compile time, then ``node.meta['val']`` is of type SymInt.
+
+For example:
+
+- ``aten::add`` returns a Tensor; so its spec will be a FakeTensor with dtype
+  and size of the tensor returned by this operator.
+- ``aten::sym_size`` returns an integer; so its val will be a SymInt because its
+  value is only available at runtime.
+- ``max_pool2d_with_indexes`` returns a tuple of (Tensor, Tensor); so the spec
+  will also be a 2-tuple of FakeTensor objects, the first TensorMeta describes
+  the first element of the return value etc.
+
+Python code::
+
+    def add_one(x):
+      return torch.ops.aten(x, 1)
+
+Graph::
+
+    graph():
+      %ph_0 : [#users=1] = placeholder[target=ph_0]
+      %add_tensor : [#users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%ph_0, 1), kwargs = {})
+      return [add_tensor]
+
+FakeTensor::
+
+    FakeTensor(dtype=torch.int, size=[2,], device=CPU)
+
+Pytree-able Types
+^^^^^^^^^^^^^^^^^
+
+We define a type “Pytree-able”, if it is either a leaf type or a container type
+that contains other Pytree-able types.
+
+Note:
+
+    The concept of pytree is the same as the one documented
+    `here <https://jax.readthedocs.io/en/latest/pytrees.html>`__ for JAX:
+
+
+The following types are defined as **leaf type**:
+
+   * - Type
+     - Definition
+   * - Tensor
+     - :class:`torch.Tensor`
+   * - Scalar
+     - Any numerical types from Python, including integral types, floating point types, and zero dimensional tensors.
+   * - int
+     - Python int (binded as int64_t in C++)
+   * - float
+     - Python float (binded as double in C++)
+   * - bool
+     - Python bool
+   * - str
+     - Python string
+   * - ScalarType
+     - :class:`torch.dtype`
+   * - Layout
+     - :class:`torch.layout`
+   * - MemoryFormat
+     - :class:`torch.memory_format`
+   * - Device
+     - :class:`torch.device`
+
+The following types are defined as **container type**:
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Type
+     - Definition
+   * - Tuple
+     - Python tuple
+   * - List
+     - Python list
+   * - Dict
+     - Python dict with Scalar keys
+   * - NamedTuple
+     - Python namedtuple
+   * - Custom class
+     - Any custom class defined with ``_register_pytree_node``
+
+Memory Formats
```

Comment:
Do we need to document this? memory format produced are up to the aten ops semantics I think.

@tugsbayasgalan @bdhirsh does aot_export enforce any particular memory format at all?

cc. @kimishpatel

Reply:
Oh yeah, we can remove it since dim_order wont be ready by PTC